feat(forms): support negative integers as indices in `FormArray`

amitbeck opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Nowadays attempting to retrieve a control in a negative index returns `undefined` (due to underlying usage of native `Array`).

Issue Number: #44642

## What is the new behavior?

Similar to the ES2021 [`Array.prototype.at()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at), support negative integers as indices:
```ts
formArray.at(-1) // equivalent to `formArray.at(formArray.length - 1)`
```

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->
Nowadays attempting to retrieve a control in a negative index returns `undefined` (due to underlying usage of native `Array`). This introduces a behavior similar to the ES2021 [`Array.prototype.at()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at).


## Other information

Consider using the native `Array.prototype.at()` when.

Closes #44642

DMezhenskyi committed:
docs: add Decoded Frontend YT channel to resources list (#44278)

PR Close #44278

iRealNirmal committed:
refactor(forms): update email validator to inherit abstractValidator (#44545)

Modified email validator to inherit abstractValidator.

For every validato type different PR will be raised as discussed in #42378.

Closes #42267

PR Close #44545

amitbeck committed:
feat(forms): support negative integers as indices in `FormArray`

amitbeck committed:
Support negative integers in `FormArray.at()`

Similar to `Array.prototype.at()`

amitbeck committed:
Support negative integers in `FormArray.at()`

Similar to `Array.prototype.at()`

amitbeck committed:
feat(forms): support negative integers as indices in `FormArray`

amitbeck committed:
Support negative integers in `FormArray.at()`

Similar to `Array.prototype.at()`

amitbeck committed:
Copy to `insert()`, `removeAt()` & `setControl()`

amitbeck committed:
Copy to `insert()`, `removeAt()` & `setControl()`

amitbeck committed:
Copy to `insert()`, `removeAt()` & `setControl()`

amitbeck committed:
Add tests

amitbeck committed:
Add tests

amitbeck committed:
Add tests

amitbeck committed:
Add tests

amitbeck committed:
Fix syntax error

amitbeck committed:
Fix syntax error

amitbeck committed:
Fix syntax error

AndrewKushnir commented:
@amitbeck thanks for creating the PR. This is an interesting proposal and I'd like to follow [the regular Feature Request process](https://github.com/angular/angular/blob/e29eb8abfbfdfb793e08ac88e14b7ba6d2ff2136/docs/GITHUB_PROCESS.md#feature-request-process) to get a signal from the community before we extend the public API surface and add extra code to the bundles. Could you please create a new feature request using the "Issues" section? Meanwhile, I'll close this PR for now and we can reopen it later if needed. Thank you.

amitbeck commented:
Thanks @AndrewKushnir, I've opened https://github.com/angular/angular/issues/44642

AndrewKushnir commented:
Reopening this PR based on https://github.com/angular/angular/issues/44642#issuecomment-1007070192.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -2050,9 +2050,11 @@ export class FormArray extends AbstractControl {
+   *     last item in the array.
    */
   at(index: number): AbstractControl {
+    index += index >= 0 ? 0 : this.length();
```
It'd be great to move this logic to a separate function:
```suggestion
    index = adjustIndex(index, this);
```
where the `adjustIndex` function may look like this:
```
function adjustIndex(index: number, formArrayInstance: FormArray): number {
  return index >= 0 ? index : index + formArrayInstance.length();
}
```

amitbeck replied:
Good idea. Would you like it to be a class method or as a util function? If it's the latter, where should I place it (in this file, in a util module)?

AndrewKushnir replied:
@amitbeck I was thinking to put it to a separate function, so that the function name can be minified (we don't do property renaming at this moment), but I guess the savings wouldn't be that big (since we'd also need to pass an argument in this case). Let's proceed with the class method.

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -42,6 +42,10 @@ describe('FormArray', () => {
 
       expect(a.controls).toEqual([c1, c3]);
+
+      a.removeAt(-1);
```
Let's add checks for all updated functions and also check situations when an index is outside of the [0, array.length - 1] range.

crisbeto committed:
fix(compiler-cli): incorrectly interpreting $any calls with a property read (#44657)

This was flagged during the code review of #44580. When generating a type check block, we were interpreting any call to `$any` as an `as any` cast, even if it's part of a `PropertyRead` (e.g. `foo.$any(1)`). This is handled correctly in other parts of the compiler, but it looks like it was missed in the type checker.

PR Close #44657

Commit:
Add tests for all updated functions

Commit:
Add tests for all updated functions

Commit:
Add tests for all updated functions

Commit:
Add tests for all updated functions

Commit:
Add and use `adjustIndex()`

Commit:
Add and use `adjustIndex()`

Commit:
Add and use `adjustIndex()`

Commit:
Merge branch 'patch-1' of https://github.com/amitbeck/angular into patch-1

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -2406,4 +2414,8 @@ export class FormArray extends AbstractControl {
     control._registerOnCollectionChange(this._onCollectionChange);
   }
+
+  private _adjustIndex(index: number): number {
```
Could you pleas add a quick comment here:
```suggestion
  /**
   * Converts negative indices to a corresponding positive number that represents
   * a position of an element in an array.
   */
  private _adjustIndex(index: number): number {
```

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -28,6 +28,16 @@ describe('FormArray', () => {
       logger = [];
     });
 
+    it('should support retrieving at an index', () => {
```
```suggestion
    it('should support retrieving an element at specified index', () => {
```

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -28,6 +28,16 @@ describe('FormArray', () => {
+      a.push(c3);
+
+      expect(a.at(1)).toEqual(c2);
+      expect(a.at(-1)).toEqual(c3);
```
Could you please also add tests for out of bounds index, like:
```
// Out of bounds index
expect(a.at(-10)).toEqual(undefined);
```

amitbeck replied:
I already did:
https://github.com/angular/angular/blob/d0246d1d0a4aee420c0c85e8f31a6b24fb982b1f/packages/forms/test/form_array_spec.ts#L38
Do you want me to add another for negative indices?

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -42,6 +52,14 @@ describe('FormArray', () => {
 
       expect(a.controls).toEqual([c1, c3]);
+
+      a.removeAt(-1);
```
Could you please also add tests for out of bounds index, like:
```
// Check that providing an out of bounds index is a noop
a.removeAt(-10);
expect(a.controls).toEqual([c1, c3]);
```

amitbeck replied:
I already did:
https://github.com/angular/angular/blob/d0246d1d0a4aee420c0c85e8f31a6b24fb982b1f/packages/forms/test/form_array_spec.ts#L60-L62
Do you want me to add another for negative indices?

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -42,6 +52,14 @@ describe('FormArray', () => {
+      
+      expect(a.controls).toEqual([c1]);
+
+      a.removeAt(a.length);
+
+      expect(a.controls).toEqual([c1]);
```
```suggestion
      // Check that using out of bounds index is a noop
      a.removeAt(a.length);
      expect(a.controls).toEqual([c1]);
```

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -65,6 +83,14 @@ describe('FormArray', () => {
 
       expect(a.controls).toEqual([c1, c2, c3]);
+
+      a.insert(-1, c2);
```
Let's create a new control (`c4`), so that it's clear where the newly-created control is inserted.

AndrewKushnir replied:
It'd also be great to add a test for out of bounds index:
```
a.insert(-10, c5);
```

amitbeck replied:
> It'd also be great to add a test for out of bounds index:
> ```
> a.insert(-10, c5);
> ```

I already did:
https://github.com/angular/angular/blob/d0246d1d0a4aee420c0c85e8f31a6b24fb982b1f/packages/forms/test/form_array_spec.ts#L91-L93
Do you want me to add another for negative indices?

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -65,6 +83,14 @@ describe('FormArray', () => {
+
+      expect(a.controls).toEqual([c1, c2, c3, c2]);
+
+      a.insert(a.length, c3);
```
Let's create a new control here as well, so that it's clear what's the position of a new element is.

amitbeck replied:
Adding `c4` and `c5`

AndrewKushnir commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -1335,6 +1361,12 @@ describe('FormArray', () => {
         expect(a.value).toEqual(['new!']);
         expect(a.valid).toBe(false);
+
+        a.setControl(-1, c);
```
Similar comment as other updated APIs:
- let's add a check for out-of-bounds index (both positive and negative ids)
- create new controls for each case (so it's clear where a new element is added)
- add some comments in the code explaining the use-case

AndrewKushnir replied:
Thanks for additional information @amitbeck. I think it still makes sense to add those checks for the `setControl` case - that'd capture the current behavior, so the test would fail in the future if we change the implementation (that'd indicate that a change is breaking).

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -2119,11 +2126,12 @@ export class FormArray extends AbstractControl {
   setControl(index: number, control: AbstractControl, options: {emitEvent?: boolean} = {}): void {
-    if (this.controls[index]) this.controls[index]._registerOnCollectionChange(() => {});
-    this.controls.splice(index, 1);
+    const adjustedIdx = this._adjustIndex(index);
```
I was looking at this change more and while having `const adjustedIdx = this._adjustIndex(index);` makes sense in a simple case (like `at`), there is a chance that both indices can be used in the code by accident. In order to avoid this, we can reuse the `index` var and set it to the adjusted index as a first line in affected functions:
```suggestion
    index = this._adjustIndex(index);
```
What do you think?

amitbeck replied:
I actually prefer your suggestion, and this is what I've done originally before implementing and using `_adjustIndex()`. I took inspiration from this code:
https://github.com/angular/angular/blob/d0246d1d0a4aee420c0c85e8f31a6b24fb982b1f/packages/core/src/linker/view_container_ref.ts#L393

AndrewKushnir committed:
ci: add golden files for runtime error codes (#44677)

Runtime error codes in the Core, Common and Forms packages were not included into the `public-api` group reviews. This commit creates the necessary golden files to keep track of further changes in the runtime codes.

This is a followup from https://github.com/angular/angular/pull/44398#issuecomment-1006910976.

PR Close #44677

amitbeck committed:
Assign to `index` instead of another variable

amitbeck committed:
Assign to `index` instead of another variable

amitbeck committed:
Add doc comment to `_adjustIndex()`

amitbeck committed:
Add doc comment to `_adjustIndex()`

amitbeck committed:
Add out-of-bounds tests for adding/removing

amitbeck committed:
Add out-of-bounds tests for adding/removing

amitbeck committed:
Move `at()` test to its own test suite

amitbeck committed:
Move `at()` test to its own test suite

amitbeck committed:
Merge remote-tracking branch 'upstream/master' into patch-1

amitbeck commented:
@alxhub @devversion @gkalpak @jelbourn @JiaLiPassion sorry for the noise, please feel free to remove yourselves as reviewers if your was only necessary due to this branch being out-of-date with `master`

thePunderWoman committed:
refactor(animations): Remove unnecessary IE specific code (#44686)

There were a few places we were still checking for Internet Explorer. This removes the references throughout the animations package.

PR Close #44686

dylhunn reviewed (approved):
Looks really nice.

dylhunn reviewed (approved):
reviewed-for: fw-forms

dylhunn commented:
It looks like you'll need to rebase on master and push again to get the CI to run.

dylhunn commented:
Oh one more thing -- in addition to rebasing, can you please squash this down to one or two logical commits? Thanks!

Commit:
Merge remote-tracking branch 'origin/patch-1' into patch-1

# Conflicts:
#	packages/forms/test/form_array_spec.ts

Commit:
support negative integers as indices in

Commit:
Merge remote-tracking branch 'origin/patch-1' into patch-1

amitbeck commented:
> Oh one more thing -- in addition to rebasing, can you please squash this down to one or two logical commits? Thanks!

@dylhunn I'm hoping ed98463 + f6d4c60 does the trick

atscott committed:
docs: release notes for the v13.1.2 release (#44699)

atscott committed:
release: cut the v13.2.0-next.2 release (#44701)

dylhunn commented:
> @dylhunn
> 
> > It looks like you'll need to rebase on master and push again to get the CI to run.
> 
> I did but it seems it still fails? 🤔

Can you tell me exactly what commands you ran? you should be able to just do `git pull upstream master && git rebase master`.

> 
> > Oh one more thing -- in addition to rebasing, can you please squash this down to one or two logical commits? Thanks!
> 
> I'm hoping [ed98463](https://github.com/angular/angular/commit/ed98463514d123301846cbc2db1cac8b56df2efc) + [f6d4c60](https://github.com/angular/angular/commit/f6d4c600e4ff318d36ed9bc3ea10d9262bd2e123) does the trick

You should be able to squash it with an interactive rebase. Do `git rebase -i HEAD~31` (since you have 31 commits), then mark the first line as `pick` and the rest as `squash`.

Thanks!

Commit:
Merge remote-tracking branch 'origin/patch-1' into patch-1

amitbeck commented:
> > @dylhunn
> > > It looks like you'll need to rebase on master and push again to get the CI to run.
> > 
> > 
> > I did but it seems it still fails? 🤔
> 
> Can you tell me exactly what commands you ran? you should be able to just do `git pull upstream master && git rebase master`.

Can't tell you for sure, I used WebStorm's "Rebase Current onto Selected" capability, but I believe it does just that.
I attempted to use `git pull upstream master && git rebase master` to rebase manually just now yet it still fails 😕

Commit:
Add out-of-bounds tests for `setControl()` and `insert()`

Commit:
Merge remote-tracking branch 'origin/patch-1' into patch-1

# Conflicts:
#	packages/forms/test/form_array_spec.ts

amitbeck commented:
> > > Oh one more thing -- in addition to rebasing, can you please squash this down to one or two logical commits? Thanks!
> > 
> > 
> > I'm hoping [ed98463](https://github.com/angular/angular/commit/ed98463514d123301846cbc2db1cac8b56df2efc) + [f6d4c60](https://github.com/angular/angular/commit/f6d4c600e4ff318d36ed9bc3ea10d9262bd2e123) does the trick
> 
> You should be able to squash it with an interactive rebase. Do `git rebase -i HEAD~31` (since you have 31 commits), then mark the first line as `pick` and the rest as `squash`.

@dylhunn Commits squashed into 06d7c2f

dylhunn commented:
Hi Amit, it looks like this PR still has 44 commit, some of which came from a bad rebase. Would you like help tidying up this history? Happy to help you get this in.

Commit:
Merge remote-tracking branch 'origin/patch-1' into patch-1

# Conflicts:
#	packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts

amitbeck commented:
> Hi Amit, it looks like this PR still has 44 commits, some of which came from a bad rebase. Would you like help tidying up the git history? Happy to help you get this in.

I'm having a hard time with the Git workflow, I won't lie. I'd love your help, please. Thanks a lot!

dylhunn commented:
> > Hi Amit, it looks like this PR still has 44 commits, some of which came from a bad rebase. Would you like help tidying up the git history? Happy to help you get this in.
> 
> I'm having a hard time with the Git workflow, I won't lie. I'd love your help, please. Thanks a lot!

Sure thing, I will rebase it later today and send a fresh PR with you as the author.

amitbeck commented:
> > > Hi Amit, it looks like this PR still has 44 commits, some of which came from a bad rebase. Would you like help tidying up the git history? Happy to help you get this in.
> > 
> > 
> > I'm having a hard time with the Git workflow, I won't lie. I'd love your help, please. Thanks a lot!
> 
> Sure thing, I will rebase it later today and send a fresh PR with you as the author.

Hey @dylhunn, did you find time to do so?

dylhunn commented:
@amitbeck, I have rebased this and sent a [new PR](https://github.com/angular/angular/pull/44746). What is your email address, for the author attribution?

dylhunn commented:
We probably need a fresh PR as commented in #44746. Thanks @amitbeck, and sorry for the difficulties with this.
